app/models/concerns/domain/lockable.rb

```diff
@@ -0,0 +1,53 @@
+module Concerns::Domain::Lockable
+  extend ActiveSupport::Concern
+
+  def apply_registry_lock
+    return unless registry_lockable?
+    return if locked_by_registrant?
+
+    transaction do
+      statuses << DomainStatus::SERVER_UPDATE_PROHIBITED
+      statuses << DomainStatus::SERVER_DELETE_PROHIBITED
+      statuses << DomainStatus::SERVER_TRANSFER_PROHIBITED
+      self.locked_by_registrant_at = Time.zone.now
+      save
+    end
+  end
+
+  def registry_lockable?
+    (statuses & [
+       DomainStatus::PENDING_DELETE_CONFIRMATION,
+       DomainStatus::PENDING_CREATE,
+       DomainStatus::PENDING_UPDATE,
+       DomainStatus::PENDING_DELETE,
+       DomainStatus::PENDING_RENEW,
+       DomainStatus::PENDING_TRANSFER,
+       DomainStatus::FORCE_DELETE,
+     ]).empty?
+  end
+
+  def locked_by_registrant?
+    return false unless locked_by_registrant_at
+
+    lock_statuses = [
+      DomainStatus::SERVER_UPDATE_PROHIBITED,
+      DomainStatus::SERVER_DELETE_PROHIBITED,
+      DomainStatus::SERVER_TRANSFER_PROHIBITED,
+    ]
+
+    (statuses & lock_statuses).count == 3
+  end
+
+  def remove_registry_lock
```

Comment:
I think `apply_registry_lock` is a bad antonym for `remove_registry_lock`. Perhaps `add` would be a better opposite for `remove`?

Reply:
Maybe `set`, but you rarely `add a lock` or `add a password`, or even `add protection`. You rather `apply` them. All are phrases from similar realm.

Reply:
I like `apply` itself. `apply/remove` pair doesn't look super, though.